src/hash/poseidon/mod.rs

```diff
@@ -0,0 +1,363 @@
+    if round_i < self.config.num_f / 2 || round_i >= self.config.num_p + self.config.num_f / 2 {
+      self.sbox_full()
+    } else {
+      self.sbox_partial()
+    }
+  }
+
+  /// applies linear layer of the round by multiplying the state with MDS matrix to mix the elements
+  fn apply_linear_layer(&mut self) {
+    let mut new_state: Vec<F> = Vec::new();
+    for i in 0..self.config.width {
+      let mut new_state_i = F::ZERO;
+      for (j, state_elem) in self.state.iter().enumerate() {
+        new_state_i += *state_elem * self.config.mds_matrix[i][j];
+      }
+      new_state.push(new_state_i);
+    }
+    self.state = new_state
+  }
+
+  /// adds round constants to the state to break the symmetricity in the state
+  fn add_round_constants(&mut self, ith: usize) {
+    for (i, state) in self.state.iter_mut().enumerate() {
+      *state += self.config.round_constants[ith * self.config.width + i];
+    }
+  }
+
+  /// perform the hashing on elements of state. state is padded by zero values, if not equal to
+  /// width of the hash function.
+  /// ## Example
+  /// ```ignore
+  /// use ronkathon::field::prime::PlutoBaseField;
+  ///
+  /// // Poseidon parameters
+  /// const WIDTH: usize = 10;
+  /// const ALPHA: usize = 5;
+  /// const NUM_P: usize = 16;
+  /// const NUM_F: usize = 8;
+  /// // load round constants and mds matrix
+  /// let (rc, mds) = load_constants::<PlutoBaseField>();
+  ///
+  /// // create a poseidon hash object with empty state
+  /// let mut poseidon = Poseidon::<PlutoBaseField>::new(WIDTH, ALPHA, NUM_P, NUM_F, rc, mds);
+  ///
+  /// // create any input
+  /// let input = std::iter::repeat(PlutoBaseField::ZERO).take(WIDTH).collect();
+  ///
+  /// let res = poseidon.hash(input);
+  /// ```
+  pub fn hash(&mut self, mut state: Vec<F>) -> F {
+    state.extend(std::iter::repeat(F::ZERO).take(self.config.width - state.len()));
+    self.state = state;
+
+    let num_rounds = self.config.num_f + self.config.num_p;
+    for i in 0..num_rounds {
+      self.add_round_constants(i);
+      self.apply_non_linear_layer(i);
+      self.apply_linear_layer();
+    }
+
+    self.state[1]
+  }
+}
+
+/// Sponge state that describes current sponge progress.
+#[derive(Clone, Copy, PartialEq)]
+pub enum SpongeState {
+  /// Sponge is initialised with state marked as zero
+  Init,
+  /// Sponge is absorbing elements
+  Absorbing,
+  /// Sponge is squeezing elements.
+  Squeezing,
+}
+
+/// Sponge config containing sponge rate, and state
+/// * `rate` describes number of elements absorbed by sponge before any permutation round
+/// * `sponge_state` describes current [`SpongeState`]. Initialised in [`SpongeState::Init`] state
+///   and, if sponge is squeezing, then it can't absorb new elements.
+/// * current new elements absorbed into hash state based on the sponge `rate`.
+pub struct SpongeConfig {
+  rate:          usize,
+  capacity:      usize,
+  sponge_state:  SpongeState,
+  absorb_index:  usize,
+  squeeze_index: usize,
+}
```

Comment:
One thing I would suggest here based on some checks you do later is use the [typestate pattern](https://zerotomastery.io/blog/rust-typestate-patterns/)

The reason why is that you can enforce that a sponge is actually a completely different *type* when it is in a different state. For instance, remove the enum and consider:

```rust
pub trait SpongeState {}

pub struct Init;
pub struct Absorbing;
pub struct Squeezing;
impl SpongeState for Init {}
impl SpongeState for Absorbing {}
impl SpongeState for Squeezing {}
```

Now, for instance, you define
```rust
pub struct PoseidonSponge<F: FiniteField, S: SpongeState> {...}
```

you definitely don't have to do this, but it would mean that you enforce that, at compile time, a sponge construction is being used properly as only specific states via a compile-time type can do specific actions.

impl<F: FiniteField> Posei

Reply:
Likewise, you can return a type that is now in a specific state too.

Reply:
I like this pattern a lot. I find it really useful for most things that have some kind of "state" to them (hence typestate as a name lol). Anything that moves an error up to compile time is always preferred imo